### Importation des libraires nécessaires

In [1]:
import pandas as pd
import numpy as np

### Importation des bases

In [2]:
dfgender = pd.read_csv(r"C:\Users\Shokh\Desktop\Formation\ML\Projet Titanic\gender_submission.csv")
test = pd.read_csv(r"C:\Users\Shokh\Desktop\Formation\ML\Projet Titanic\test.csv")
train = pd.read_csv(r"C:\Users\Shokh\Desktop\Formation\ML\Projet Titanic\train.csv")

### Exploration du dataset

In [87]:
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Valeurs manquantes ?

In [88]:
train.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

177 valeurs manquantes dans la colonne 'Âge' (importante) et 687 dans la colonne "Cabin" (moins importante pour notre analyse)

### Description des colonnes (features) du dataset :

**PassengerID :** L'ID (unique) du passager, l'index.

**Survived :** Indique la survie ou non du passager, valeur binaire 0 (non survécu) ou 1 (survécu) type INT

**Pclass :** Classe du ticket du passager, 1ère (1), 2ème (2), 3ème (3) type INT

**Name :** Nom du passager, type STRING.

**Sex :** Sexe du passager, homme ou femme type STRING

**Age :** Âge du passager type INT

**SibSp :** Nombre de frères/soeurs ou conjoints du passager à bord

**Parch :** Nombre de parents/enfants du passager à bord

**Ticket :** Numéro du ticket

**Fare :** Prix du billet du passager

**Cabin :** Numéro de la cabine du passager

**Embarked :** Port d'embarcation du passager : C = Cherbourg, Q = Queenstown, S = Southampton

In [29]:
train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

Pour cette commande, ici les seules colonnes pertinentes sont l'âge et le fare (prix du billet)

In [3]:
train = train.drop(columns=['Cabin','Ticket','Name'])

In [90]:
train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch',
       'Fare', 'Embarked'],
      dtype='object')

In [91]:
train.shape

(891, 9)

In [92]:
train.head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0            1         0       3    male  22.0      1      0   7.2500        S
1            2         1       1  female  38.0      1      0  71.2833        C
2            3         1       3  female  26.0      0      0   7.9250        S
3            4         1       1  female  35.0      1      0  53.1000        S
4            5         0       3    male  35.0      0      0   8.0500        S

In [4]:
train['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [5]:
train['Embarked'].fillna('S', inplace=True)

In [95]:
train.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Sex              0
Age            177
SibSp            0
Parch            0
Fare             0
Embarked         0
dtype: int64

In [6]:
train['Age'].fillna(15,limit=59,inplace=True)
train['Age'].fillna(30,limit=59,inplace=True)

In [7]:
train.isna().sum()

PassengerId     0
Survived        0
Pclass          0
Sex             0
Age            59
SibSp           0
Parch           0
Fare            0
Embarked        0
dtype: int64

In [8]:
train['Age'].fillna(45,inplace=True)

In [9]:
train.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Sex            0
Age            0
SibSp          0
Parch          0
Fare           0
Embarked       0
dtype: int64

Plus aucune valeur manquantes. Il y avait 177 valeurs manquantes dans la colonne 'Age', nous avons regardé la moyenne et l'écart-type sachant que nous avons une moyenne de 30 et un écart type de 15. Ensuite nous avons divisé le nombre de valeurs manquantes par 3 pour les distribuer de façon la plus équitable possible. Donc 59 modalités par 30 (moyenne) ensuite 59 modalités par 15 (30-écart type = 15) puis 59 modalités par 45 (moyenne + écart type = 45)

In [10]:
corr_matrix = train.corr()
corr_matrix.style.background_gradient(cmap='coolwarm') # Aucune feature n'est vraiment corrélé à resilies

In [101]:
train.to_csv(r'D:\Git\Brief_Projet_Titanic\titanic\clean_train.csv')

In [102]:
train.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Sex             object
Age            float64
SibSp            int64
Parch            int64
Fare           float64
Embarked        object
dtype: object

# Regression linéaire

#### Importation librairies pour régression linéaire

In [79]:
from sklearn.linear_model import *
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.metrics import *

In [12]:
model = LinearRegression()

In [13]:
train.head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0            1         0       3    male  22.0      1      0   7.2500        S
1            2         1       1  female  38.0      1      0  71.2833        C
2            3         1       3  female  26.0      0      0   7.9250        S
3            4         1       1  female  35.0      1      0  53.1000        S
4            5         0       3    male  35.0      0      0   8.0500        S

### Passage du qualitative au quantitatif pour la régression linéaire

In [14]:
train.loc[train["Sex"] == "male", "Sex"] = 0
train.loc[train["Sex"] == "female", "Sex"] = 1

In [15]:
train.loc[train["Embarked"] == "S", "Embarked"] = 0
train.loc[train["Embarked"] == "C", "Embarked"] = 1
train.loc[train["Embarked"] == "Q", "Embarked"] = 2

## Prédiction avec les différentes features

Avec **l'âge** comme seule feature

In [43]:
X = train[['Age']]
y = train[['Survived']]

In [44]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=3)

In [45]:
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [46]:
predictions = model.predict(X_test)

In [47]:
score = model.score(X,y)
score

0.006604778290112301

Avec pour seule feature l'âge, le score est extrêmement mauvais en régression linéaire.

Avec le **sexe** comme seule feature

In [48]:
X = train[['Sex']]
y = train[['Survived']]

In [49]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=5)

In [50]:
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [51]:
predictions = model.predict(X_test)

In [52]:
score = model.score(X_test,y_test)
score

0.3035054335474724

In [53]:
X = train[['Pclass']]
y = train[['Survived']]

In [54]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=5)

In [55]:
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [56]:
predictions = model.predict(X_test)

In [57]:
score = model.score(X_test,y_test)
score

0.14617885248461615

In [58]:
X = train[['Fare']]
y = train[['Survived']]

In [59]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=5)

In [60]:
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [61]:
predictions = model.predict(X_test)

In [63]:
score = model.score(X_test,y_test)
score

0.05300941421809047

In [64]:
X = train[['Sex','Age','Pclass','Fare']]
y = train[['Survived']]

In [65]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=5)

In [66]:
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [67]:
predictions = model.predict(X_test)

In [68]:
score = model.score(X_test,y_test)
score

0.3697808481854563

Le meilleur score s'obtient donc quand on met toutes les features, Age Sex Class et Fare. Score bien meilleure qu'avec une seule de ces variable prises individuellement mais toujours peu convaincant, nous ne sommes qu'à 0.37.

In [73]:
X = train[['Sex','Age','Pclass']]
y = train[['Survived']]

In [74]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=None)

In [75]:
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [76]:
predictions = model.predict(X_test)

In [77]:
score = model.score(X_test,y_test)
score

0.36304398326222953

In [80]:
sklearn.metrics.r2_score(y_test,predictions)

NameError: name 'sklearn' is not defined